torch/testing/_internal/common_utils.py

```diff
@@ -691,12 +689,121 @@ def read_and_process():
             "Test is disabled because an issue exists disabling it: {}".format(disabled_test_from_issues[test_name]) +
             " To enable set the environment variable PYTORCH_RUN_DISABLED_TESTS=1")
 
+# Helper that returns the maximum value a dtype can represent
+def _get_width(dtype):
+    if dtype is torch.complex128:
+        return torch.finfo(torch.float64).bits
+    elif dtype is torch.complex64:
+        return torch.finfo(torch.float32).bits
+    elif dtype is torch.bool:
+        return 1
+    elif is_integral(dtype):
+        return torch.iinfo(dtype).bits
+    return torch.finfo(dtype).bits
+
+# Returns a dtype suitable for comparing tensors a and b.
+# NOTE: Dtypes for Comparing Tensors in the Test Framework
+# When two tensors are compared they may be of different dtypes.
+# In PyTorch, the typical type promotion rules are biased towards float
+# types, so tensors with dtypes float16 and int64 produce tensors of
+# dtype float16 when added together.
+# If we used typical PyTorch type promotion to determine the dtype that tensors
+# should be compared in, however, then the test framework would assert
+# that tensors which are actually very different were "equal" to each other.
+# For example, torch.tensor((50000), dtype=torch.long) would be "equal" to
+# torch.tensor(49984., dtype=torch.float16), regardless of the rtol and atol.
+# These cases are admittedly, unlikely, but the test framework is not intended
+# to reflect only "likely" scenarios.
+# To preserve numerical accuracy, instead of following PyTorch's type
+# promotion rules this function follows NumPy's, which are intended to preserve
+# numerics. It special-cases bfloat16 since NumPy doesn't support bfloat16.
+# TODO: if we take a NumPy dependency then this function can be greatly
+# simplified.
```

Comment:
I don't know if we'd add a numpy dependency, though when we do add a subdtype api, maybe that would help make the `kind` specification simpler

Reply:
True. Could be a few ways to simplify.